src/helpers/connection.ts

```diff
@@ -0,0 +1,45 @@
+import { type ChainName, getBaseUrl } from "./chains.js";
+import { type Signer, type ExternalProvider, getSigner } from "./ethers.js";
+
+export interface ReadConfig {
+  baseUrl: string;
+}
+
+export interface SignerConfig {
+  signer: Signer;
+}
+
+export interface AutoWaitConfig {
+  autoWait: boolean;
+}
+
+export type Config = Partial<ReadConfig & SignerConfig>;
+
+export async function extractBaseUrl(
```

Comment:
This function does its best to figure out where to route queries based on defaults, if you have a signer connected to a chain, if you have a fallback chain you want to use, etc.